velox/type/TypeUtil.cpp

```diff
@@ -33,4 +33,22 @@ velox::RowTypePtr concatRowTypes(
   return velox::ROW(std::move(columnNames), std::move(columnTypes));
 }
 
+const velox::TypePtr isHomogeneousRow(const velox::TypePtr& type) {
+  if (!type || type->kind() != velox::TypeKind::ROW) {
+    return nullptr;
+  }
+  auto row = velox::asRowType(type);
```

Comment:
const auto& to avoid copy

However, we can avoid this cast altogether by using size() and childAt(index) APIs.